### ISSUE TITLE
Using the Web Share API in React

### ISSUE THREAD

**Issue by earthlingdavey** [cbrannen9a/brannen-dev#29]
Thanks for your article at https://www.brannen.dev/posts/using-the-web-share-api-in-react

It's really helpful, and I learned a lot, not just of the API but some useful TS stuff too.

I'll soon be using it in production for my startup.

I noticed a couple of micro improvements that you may like:

You could use `?.` syntax to ensure a function existss before running it.

e.g. `onInteraction?.();` can replace `onInteraction && onInteraction();` 

I spotted `WithColors` type in step 4, I guess it's an artifact from your codebase?

Additional feature: could onSuccess be triggered by copy, and also return the method of sharing?

Strating by modifying the type to `onSuccess?: (method: 'native' | 'copy') => void;` and passing that function to the `SharePopup` component.

Lastly, and probably the least important, I added support for ancient browsers with an extra few lines, I may as well also share it here.

```ts
// ShareController.tsx

if (navigator?.share) {
```

```ts
// SharePopup.tsx

import { type FC, type MouseEvent, useRef, useState } from 'react';

// ...

const textElementRef = useRef<null | HTMLInputElement>(null);

const copyClicked = async (
  e: MouseEvent<HTMLButtonElement, globalThis.MouseEvent>
) => {
  try {
    // clipboard.writeText has wide but not 100% support
    // https://caniuse.com/?search=writeText
    await navigator.clipboard.writeText(shareData?.url || '');
    setState('success');
  } catch (err1) {
    try {
      // execCommand has > 97% support but is deprecated, use it as a fallback
      // https://caniuse.com/?search=execCommand
      if (!textElementRef.current) {
        throw new Error('textElementRef.current is null');
      }
      textElementRef.current.select();
      document.execCommand('copy');
      e.currentTarget.focus();
      setState('success');
    } catch (err2) {
      onError?.([err1, err2]);
      setState('error');
    }
  }
};

// ...

<input ref={textElementRef} value={shareData.url} readOnly />
```

Cheers again for the informative article, I hope my feedback was well received. ✌️

**Comment by cbrannen9a** [comment]
Thank you for the feedback, much appreciated. 

I'll look at updating the article with the points you raised.

Glad it helped you 👍

**Comment by earthlingdavey** [comment]
Yes absolutely it helped. The button is live now on our [list template](https://usedandloved.com/items/filters/list:(RA)/) - in case you wanted to check it. Cheers :)
